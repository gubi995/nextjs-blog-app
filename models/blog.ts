export interface Blog {
  id: string;
  title: string;
  createdAt: string;
  introduction: string;
  content: string;
  slug: string;
  image: {
    url: string;
  };
}
