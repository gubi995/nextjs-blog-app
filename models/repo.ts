export interface Repo {
  createdAt: string;
  name: string;
  homepageUrl: string;
  url: string;
}
