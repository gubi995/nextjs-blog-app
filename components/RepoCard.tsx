import React from 'react';
import { Paper, Typography, Divider, styled, Button } from '@material-ui/core';

import { Repo } from '../models/repo';
import { toReadableDate } from '../utils/date';

const StyledRepoCard = styled(Paper)(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  padding: theme.spacing(2),
  '& > *': {
    marginBottom: theme.spacing(1),
    marginTop: theme.spacing(1),
  },
}));

const StyledContainer = styled('div')({ display: 'flex', justifyContent: 'space-between' });

const Info = ({ caption, value }: { caption: string; value: string }) => {
  return (
    <div>
      <Typography variant="h6" color="primary">
        {caption}
      </Typography>
      <Typography variant="overline" color="primary">
        {value}
      </Typography>
    </div>
  );
};

interface Props {
  repo: Repo;
}

const RepoCard = ({ repo }: Props) => {
  return (
    <StyledRepoCard>
      <StyledContainer>
        <Info caption="Name" value={repo.name} />
        <Info caption="Created" value={toReadableDate(repo.createdAt)} />
      </StyledContainer>

      <Divider orientation="horizontal" color="primary" />

      <StyledContainer>
        <Button href={repo.url ? repo.url : '#'} rel="noopener" target="_blank" color="primary">
          Repo url
        </Button>
        <Divider orientation="vertical" />
        <Button
          href={repo.homepageUrl ? repo.homepageUrl : '#'}
          rel="noopener"
          target="_blank"
          disabled={!Boolean(repo.homepageUrl)}
          color="primary"
        >
          Home page url
        </Button>
      </StyledContainer>
    </StyledRepoCard>
  );
};

export default RepoCard;
