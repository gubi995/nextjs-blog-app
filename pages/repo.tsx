import { NextPage, GetStaticProps } from 'next';
import { Container, styled } from '@material-ui/core';
import { GraphQLClient } from 'graphql-request';

import { Repo } from '../models/repo';
import { GET_PERSONAL_REPOS } from '../graphql/queries';
import { getHeaders } from '../utils/graphql-client';
import AppShell from '../components/AppShell';
import CustomBanner from '../components/CustomBanner';
import RepoCard from '../components/RepoCard';

const StyledContainer = styled(Container)(({ theme }) => ({ '& > *': { margin: theme.spacing(4) } }));

interface Props {
  repositories: Repo[];
}

const RepoPage: NextPage<Props> = ({ repositories }) => {
  return (
    <AppShell pageTitle="Repositories page">
      <CustomBanner title="Repositories ðŸŽ®" />
      <StyledContainer maxWidth="md">
        {repositories.map((repo) => (
          <RepoCard repo={repo} key={repo.url} />
        ))}
      </StyledContainer>
    </AppShell>
  );
};

export const getStaticProps: GetStaticProps = async (context) => {
  const client = new GraphQLClient('https://api.github.com/graphql', getHeaders(process.env.GITHUB_TOKEN!));

  const data = await client.request(GET_PERSONAL_REPOS);

  const repositories = data.viewer.repositories.nodes;

  return {
    props: { repositories },
  };
};

export default RepoPage;
