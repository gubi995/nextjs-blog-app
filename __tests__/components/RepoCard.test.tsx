import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import RepoCard from '../../components/RepoCard';
import { Repo } from '../../models/repo';

const repo: Repo = {
  createdAt: '2020-04-01T07:14:32Z',
  homepageUrl: 'http://mock-home-page.com',
  url: 'http://mock-repo.com',
  name: 'Mock repo',
};

const repoWithoutUrls: Repo = { ...repo, url: undefined, homepageUrl: undefined };

describe('RepoCard', () => {
  test("should rendered component matches with it's the snapshot [URLs are present]", () => {
    const { asFragment } = render(<RepoCard repo={repo} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("should rendered component matches with it's the snapshot [URLs are NOT present]", () => {
    const { asFragment } = render(<RepoCard repo={repoWithoutUrls} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should have "href" attributes have been set to "a" tags', () => {
    const { getByText } = render(<RepoCard repo={repo} />);

    const repoBtn = getByText('Repo url').closest('a');
    const homeBtn = getByText('Home page url').closest('a');

    expect(repoBtn).toHaveAttribute('href', 'http://mock-repo.com');
    expect(homeBtn).toHaveAttribute('href', 'http://mock-home-page.com');
    expect(homeBtn).toHaveAttribute('aria-disabled', 'false');
  });

  test('should NOT have "href" attributes have been set to "a" tags', () => {
    const { getByText, debug } = render(<RepoCard repo={repoWithoutUrls} />);

    const repoBtn = getByText('Repo url').closest('a');
    const homeBtn = getByText('Home page url').closest('a');

    console.log(debug());

    expect(repoBtn).toHaveAttribute('href', '#');
    expect(homeBtn).toHaveAttribute('href', '#');
    expect(homeBtn).toHaveAttribute('aria-disabled', 'true');
  });
});
