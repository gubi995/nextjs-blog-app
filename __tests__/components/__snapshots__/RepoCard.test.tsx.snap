// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoCard should rendered component matches with it's the snapshot [URLs are NOT present] 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root WithStyles(ForwardRef(Paper))-root-111 WithStyles(ForwardRef(Paper))-root-112 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="div-root-141"
    >
      <div>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary"
        >
          Name
        </h6>
        <span
          class="MuiTypography-root MuiTypography-overline MuiTypography-colorPrimary"
        >
          Mock repo
        </span>
      </div>
      <div>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary"
        >
          Created
        </h6>
        <span
          class="MuiTypography-root MuiTypography-overline MuiTypography-colorPrimary"
        >
          4/1/2020
        </span>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
      color="primary"
    />
    <div
      class="div-root-141"
    >
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
        href="#"
        rel="noopener"
        tabindex="0"
        target="_blank"
      >
        <span
          class="MuiButton-label"
        >
          Repo url
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
      <hr
        class="MuiDivider-root MuiDivider-vertical"
      />
      <a
        aria-disabled="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
        href="#"
        rel="noopener"
        tabindex="-1"
        target="_blank"
      >
        <span
          class="MuiButton-label"
        >
          Home page url
        </span>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RepoCard should rendered component matches with it's the snapshot [URLs are present] 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root WithStyles(ForwardRef(Paper))-root-1 WithStyles(ForwardRef(Paper))-root-2 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="div-root-31"
    >
      <div>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary"
        >
          Name
        </h6>
        <span
          class="MuiTypography-root MuiTypography-overline MuiTypography-colorPrimary"
        >
          Mock repo
        </span>
      </div>
      <div>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary"
        >
          Created
        </h6>
        <span
          class="MuiTypography-root MuiTypography-overline MuiTypography-colorPrimary"
        >
          4/1/2020
        </span>
      </div>
    </div>
    <hr
      class="MuiDivider-root"
      color="primary"
    />
    <div
      class="div-root-31"
    >
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
        href="http://mock-repo.com"
        rel="noopener"
        tabindex="0"
        target="_blank"
      >
        <span
          class="MuiButton-label"
        >
          Repo url
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
      <hr
        class="MuiDivider-root MuiDivider-vertical"
      />
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
        href="http://mock-home-page.com"
        rel="noopener"
        tabindex="0"
        target="_blank"
      >
        <span
          class="MuiButton-label"
        >
          Home page url
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;
